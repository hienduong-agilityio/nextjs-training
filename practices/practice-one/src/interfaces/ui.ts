export interface IInputGroupProps {
  placeholder?: string;
  value?: string;
  buttonText?: string;
  inputName?: string;
  isDisabled?: boolean;
  isLoading?: boolean;
  startIcon?: React.ReactNode;
  customClass?: {
    container?: string;
    inputContainer?: string;
    input?: string;
    button?: string;
  };
  onSubmit?: (event: React.FormEvent<HTMLFormElement>) => void;
}
