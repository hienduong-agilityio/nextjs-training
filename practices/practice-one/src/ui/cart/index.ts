export * from './CartTable';
export * from './OrderSummary';
export * from './VoucherInput';
export * from './EmptyCart';

export { default as CartItemRow } from './CartItemRow';
