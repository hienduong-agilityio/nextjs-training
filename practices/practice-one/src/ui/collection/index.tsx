export * from './LoadingCategoryGroup';
export * from './LoadingFilterSortBar';
