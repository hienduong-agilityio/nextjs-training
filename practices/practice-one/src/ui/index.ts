export * from './products/index';
export * from './collection/index';
