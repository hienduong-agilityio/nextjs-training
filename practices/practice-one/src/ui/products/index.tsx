export * from './CartAndFavoriteActions';
export * from './ProductImages';
export * from './ProductInfo';
export * from './QuantityControl';
export * from './SocialShare';
export * from './RelatedProduct';
export * from './ProductDetailTabs';
export * from './ProductReview';
export * from './ProductActions';
export * from './ProductLoading';
