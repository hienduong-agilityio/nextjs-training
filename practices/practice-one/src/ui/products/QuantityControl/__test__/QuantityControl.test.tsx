import { render, screen, fireEvent } from '@testing-library/react';

// UI
import { QuantityControl } from '@/ui';

describe('QuantityControl Component', () => {
  const renderQuantityControl = (props = {}) => {
    return render(
      <QuantityControl initialQuantity={1} maxQuantity={10} {...props} />,
    );
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('matches the snapshots', () => {
    const { asFragment } = renderQuantityControl();
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders with default props', () => {
    renderQuantityControl();

    const input = screen.getByRole('spinbutton');
    expect(input).toBeInTheDocument();
    expect(input).toHaveValue(1);

    const decrementButton = screen.getByText('-');
    const incrementButton = screen.getByText('+');

    expect(decrementButton).toBeInTheDocument();
    expect(incrementButton).toBeInTheDocument();
  });

  it('increments the quantity when increment button is clicked', () => {
    const onQuantityChange = jest.fn();
    renderQuantityControl({ onQuantityChange });

    const incrementButton = screen.getByText('+');
    fireEvent.click(incrementButton);

    const input = screen.getByRole('spinbutton');
    expect(input).toHaveValue(2);
    expect(onQuantityChange).toHaveBeenCalledWith(2);
  });

  it('decrements the quantity when decrement button is clicked', () => {
    const onQuantityChange = jest.fn();
    renderQuantityControl({ initialQuantity: 2, onQuantityChange });

    const decrementButton = screen.getByText('-');
    fireEvent.click(decrementButton);

    const input = screen.getByRole('spinbutton');
    expect(input).toHaveValue(1);
    expect(onQuantityChange).toHaveBeenCalledWith(1);
  });

  it('disables decrement button when quantity is at minimum', () => {
    renderQuantityControl({ initialQuantity: 1 });

    const decrementButton = screen.getByText('-');
    expect(decrementButton).toBeDisabled();
  });

  it('disables increment button when quantity is at maximum', () => {
    renderQuantityControl({ initialQuantity: 10, maxQuantity: 10 });

    const incrementButton = screen.getByText('+');
    expect(incrementButton).toBeDisabled();
  });

  it('clamps the quantity to max value when input exceeds max quantity', () => {
    const onQuantityChange = jest.fn();
    renderQuantityControl({ maxQuantity: 10, onQuantityChange });

    const input = screen.getByRole('spinbutton');
    fireEvent.change(input, { target: { value: '15' } });
    fireEvent.blur(input);

    expect(input).toHaveValue(10);
    expect(onQuantityChange).toHaveBeenCalledWith(10);
  });

  it('clamps the quantity to minimum value when input is below 1', () => {
    const onQuantityChange = jest.fn();
    renderQuantityControl({ onQuantityChange });

    const input = screen.getByRole('spinbutton');
    fireEvent.change(input, { target: { value: '-3' } });
    fireEvent.blur(input);

    expect(input).toHaveValue(1);
    expect(onQuantityChange).toHaveBeenCalledWith(1);
  });

  it('disables all controls when isLoading is true', () => {
    renderQuantityControl({ isLoading: true });

    const decrementButton = screen.getByText('-');
    const input = screen.getByRole('spinbutton');

    expect(decrementButton).toBeDisabled();
    expect(input).toBeDisabled();
  });
});
