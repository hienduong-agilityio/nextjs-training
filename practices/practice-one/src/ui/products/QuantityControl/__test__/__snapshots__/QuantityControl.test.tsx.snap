// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuantityControl Component matches the snapshots 1`] = `
<DocumentFragment>
  <div
    aria-label="Quantity Control"
    class="rounded-md gap-4 bg-secondary-50"
  >
    <div
      class="flex"
    >
      <button
        aria-label="decrement"
        class="px-5 py-3 text-primary-100 hover:text-gray-800 disabled:cursor-not-allowed disabled:text-gray-800"
        disabled=""
        type="button"
      >
        -
      </button>
      <input
        aria-label="Quantity input"
        class="w-12 text-center border-0 bg-gray-50 focus:ring-0"
        type="number"
        value="1"
      />
      <button
        aria-label="increment"
        class="px-5 py-3 text-primary-100 hover:text-gray-800 disabled:cursor-not-allowed disabled:text-gray-800"
        type="button"
      >
        +
      </button>
    </div>
  </div>
</DocumentFragment>
`;
